Add SsmGetCommandInvocationOperator and enhance SSM components

ksharlandjiev opened the pull request:
## Description

This PR adds a new `SsmGetCommandInvocationOperator` and enhances existing SSM components to provide better command output retrieval capabilities.

## Changes

### New Features
- **SsmGetCommandInvocationOperator**: New operator to retrieve output and execution details from SSM command invocations
  - Supports retrieving output from specific instances or all instances
  - Returns structured data with stdout, stderr, execution times, and status
  - Handles errors gracefully with proper logging

### Enhancements
- **SsmHook**: Added `get_command_invocation()` and `list_command_invocations()` methods
- **SsmRunCommandOperator**: Enhanced deferrable mode with proper AWS parameter passing to triggers
- **SsmRunCommandTrigger**: Improved async execution and AWS connection parameter handling

### Testing
- Added unit tests for all new functionality
- Updated the system test demonstrating end-to-end usage

### Documentation
- Updated RST documentation with usage examples

## Use Cases

This enhancement enables several important use cases:

1. **Command Output Retrieval**: Get stdout/stderr from SSM commands for downstream processing
2. **Debugging**: Inspect command execution details for troubleshooting
3. **Data Processing**: Use SSM command output as input for subsequent tasks
4. **Monitoring**: Track command execution status and performance metrics

## Example Usage

```python
# Run a command
run_command = SsmRunCommandOperator(
    task_id="run_command",
    document_name="AWS-RunShellScript",
    run_command_kwargs={
        "InstanceIds": ["i-1234567890abcdef0"],
        "Parameters": {"commands": ["echo 'Hello World'"]}
    }
)

# Get the output
get_output = SsmGetCommandInvocationOperator(
    task_id="get_output",
    command_id="{{ ti.xcom_pull(task_ids='run_command') }}",
    instance_id="i-1234567890abcdef0"
)

Commit:
initial commit

Commit:
Add SsmGetCommandInvocationOperator for retrieving SSM command output

This operator allows users to retrieve the output and execution details
from SSM commands, supporting both single instance and all instances scenarios.

Features:
- Retrieve stdout/stderr from SSM command executions
- Support for specific instance or all instances
- Structured output with execution metadata
- Comprehensive error handling

Includes unit tests, documentation updates, and system test integration.

Commit:
Pass AWS connection parameters to SsmRunCommandTrigger in deferrable mode

Fixed deferrable SSM commands failing due to missing region configuration:
  * Fix NoRegionError in SsmRunCommandOperator when using deferrable=True
  * Add region_name, verify, and botocore_config parameters to SsmRunCommandTrigger constructor
  * Update SsmRunCommandOperator to pass AWS connection parameters to trigger
  * Add unit tests for trigger serialization and parameter passing

The SsmRunCommandOperator was failing in deferrable mode with 'NoRegionError: You must specify a region'
because the trigger was not receiving AWS connection parameters from the operator. This change ensures
all AWS configuration (region_name, verify, botocore_config) is properly passed to the trigger and
maintains backward compatibility.

Commit:
standartise SsmGetCommandInvocationOperator output format.

- Ensure consistent return format.
- Update tests to match standardised output structure

The operator now consistently returns the same structure regardless of whether
instance_id is provided, making it easier to work with downstream tasks.

Commit:
Standardise SSM hook return format to match AWS patterns

- Change SsmHook.list_command_invocations() to return raw AWS API response
- Update return type from list[dict] to dict for consistency with other AWS hooks
- Modify SsmGetCommandInvocationOperator to handle raw response format
- Update all unit tests to expect standardized response format

This aligns SSM hook with the established pattern used by EC2Hook, LambdaHook,
and other AWS hooks that return raw boto3 responses rather than processed data."

Commit:
Fixing hooks unit test to use mock decorator and valid UUID

Commit:
final touches on the operator

Commit:
Updated exception handling to use `RuntimeError` instead of `AirflowException` per latest contribution guidelines

Commit:
Fix test_poke_failure_states to expect RuntimeError vs AirflowException (as per latest community guidelines)

Commit:
fix: doc spelling mistake

ksharlandjiev committed:
Merge branch 'apache:main' into feature/ssm-get-command-invocation-operator

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/operators/ssm.py:
```diff
@@ -129,3 +132,91 @@ def execute(self, context: Context):
+            invocations = response.get("CommandInvocations", [])
+
+        # TBD: Do we return formatted or RAW API output? 
+        # My rationalle for going formatted: Raw AWS API response contains a lot of metadata that's typically not needed for downstream tasks.
```
I prefer the operators to be a simplified veil in front of the the raw boto api. So I'd also agree with a formatted response that contains less than the full response.

ferruzzi replied:
I can get behind this.  Feel free to remove the TBD comment and resolve this conversation.

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/operators/ssm.py:
```diff
@@ -90,7 +90,7 @@ def execute_complete(self, context: Context, event: dict[str, Any] | None = None
 
         if event["status"] != "success":
-            raise AirflowException(f"Error while running run command: {event}")
+            raise RuntimeError(f"Error while running run command: {event}")
```
Was there any other code that was catching the previous AriflowException that now needs to be updated to catch the RuntimeException?

ksharlandjiev replied:
Yes, there was a unit test, which blew up, and I've fixed here: https://github.com/apache/airflow/pull/56936/commits/26263773625951e6dfb44685de8955c0e4e9a260 (first commit after the PR)

o-nikolas replied:
But no source code? I find it odd that we were raising a specific exception but nothing was catching it but a unit test?

ksharlandjiev replied:
@o-nikolas I've checked the entire repository. The only place I've found is the SsmRunCommandTrigger class still raises AirflowException in its run() method. 

I'm happy to revert this and use AirflowException, if this feels safer from backward compatibility perspective.

o-nikolas replied:
It's all good, if you've verified nothing was specifically catching that exception then this is a welcomed change since we're trying to deprecate AirflowException :)

o-nikolas reviewed (commented):
Thanks for the contribution! Overall it looks quite good. Some feedback:

I don't love the mixture of changes. It's better to make your PRs as tightly scoped as possible, if you spot other things to fix try to do them in a separate PR.

You have some tests failing that need looking into, static checks can be run on your development machine (automatically as well as pre-commit/pre-push hooks), see the docs [here](https://github.com/apache/airflow/blob/main/contributing-docs/08_static_code_checks.rst) for how to run those.

Commit:
fix: resolve E501 line length violations in SSM operators

- Break long lines in SsmRunCommandOperator and SsmGetCommandInvocationOperator
- Split long method calls, dictionary configurations, and log messages
- Ensure all lines comply with 79-character limit per PEP 8

Commit:
Fix RST formatting in SsmRunCommandCompletedSensor docstring

Commit:
ruff fix

ferruzzi commented on providers/amazon/docs/operators/ssm.rst:
```diff
@@ -58,6 +58,43 @@ or the :class:`~airflow.providers.amazon.aws.triggers.ssm.SsmRunCommandTrigger`
+
+This operator is useful for:
+
+* Retrieving output from commands executed by ``SsmRunCommandOperator`` in previous tasks
```
Nitpick:  Consider using the :class: format here like you did above (currently line 67) to make it a convenient link.  Or feel free to just resolve this, I'm not fussed.

ferruzzi reviewed (commented):
Looking good.

Commit:
fix(amazon): Fix SSM documentation build errors and improve formatting. Thank you @ferruzzi for your guidance.

Commit:
Remove TBD comment in SSM operator

ferruzzi reviewed (approved):
Nice, thanks for your work!
